server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CachedBlobStore.java

```diff
@@ -0,0 +1,167 @@
+
+import reactor.core.publisher.Mono;
+
+public class CachedBlobStore implements BlobStore {
+
+    private static final String DEFAULT_BUCKET = "cassandraDefault";
+
+    private final BlobStoreCache cache;
+    private final BlobStore backend;
+    private final Integer sizeThresholdInBytes;
+    private final BucketName defaultBucket;
+
+    @Inject
+    public CachedBlobStore(BlobStoreCache cache, BlobStore backend,
+                           CassandraCacheConfiguration cacheConfiguration,
+                           @Named(DEFAULT_BUCKET) BucketName defaultBucket) {
+        this.cache = cache;
+        this.backend = backend;
+        this.sizeThresholdInBytes = cacheConfiguration.getSizeThresholdInBytes();
+        this.defaultBucket = defaultBucket;
+    }
+
+    @Override
+    public InputStream read(BucketName bucketName, BlobId blobId) throws ObjectStoreIOException, ObjectNotFoundException {
+        Preconditions.checkNotNull(bucketName, "bucketName should not be null");
+
+        return Mono.just(bucketName)
+            .filter(defaultBucket::equals)
+            .flatMap(ignored ->
+                Mono.from(cache.read(blobId))
+                    .<InputStream>flatMap(bytes -> Mono.fromCallable(() -> new ByteArrayInputStream(bytes))))
+            .switchIfEmpty(Mono.fromCallable(() -> backend.read(bucketName, blobId)))
+            .blockOptional()
+            .orElseThrow(() -> new ObjectNotFoundException(String.format("Could not retrieve blob metadata for %s", blobId)));
+    }
+
+    @Override
+    public Mono<byte[]> readBytes(BucketName bucketName, BlobId blobId) {
+        return Mono.just(bucketName)
+            .filter(defaultBucket::equals)
+            .flatMap(ignored -> Mono.from(cache.read(blobId)))
+            .switchIfEmpty(Mono.from(backend.readBytes(bucketName, blobId)));
+    }
+
+    @Override
+    public Mono<BlobId> save(BucketName bucketName, byte[] bytes, StoragePolicy storagePolicy) {
+        return Mono.from(backend.save(bucketName, bytes, storagePolicy))
+            .flatMap(blobId -> {
+                if (isAbleToCache(bucketName, bytes, storagePolicy)) {
+                    return Mono.from(cache.cache(blobId, bytes)).thenReturn(blobId);
+                }
+                return Mono.just(blobId);
+            });
+    }
+
+    @Override
+    public Publisher<BlobId> save(BucketName bucketName, InputStream inputStream, StoragePolicy storagePolicy) {
+        Preconditions.checkNotNull(inputStream, "InputStream must not be null");
+
+        if (isAbleToCache(bucketName, storagePolicy)) {
+            return saveInCache(bucketName, inputStream, storagePolicy);
+        }
+
+        return backend.save(bucketName, inputStream, storagePolicy);
+    }
+
+    private Publisher<BlobId> saveInCache(BucketName bucketName, InputStream inputStream, StoragePolicy storagePolicy) {
+        PushbackInputStream pushbackInputStream = new PushbackInputStream(inputStream, sizeThresholdInBytes + 1);
```

Comment:
this pushback inputstream introduce a strong dependency with `isAbleToCache()`. Could we instantiate it in `isAbleToCache()` and return it here?

Reply:
I think it not.
`isAbleToCache()` is like yes/no question so it should return true/false.BTW  this method are common use for others method too.

Reply:
the problem is: depending of how the PushbackInputStream is configured, the code will work or not, it's what we call "temporal coupling".
If we can't make it work with method extractions, it probably means we need a type to encapsulate that?

Reply:
All the classes doing such things all rely (locally) on temporal coupling.

We live with it in the past, and it is located in a single class. Sounds okyish to me.

Otherwize we can introduce a conpound type:

```
static class {
    private final PushBackInputStream stream;
    private final int readableSize;
}
```

However, that will not even solve your problem, as the stream might have been pre-consumend, invalidating the readableSize assumption. Ahhh mutability!

I advocate a comment about the (temporal coupled) requirements the consuming method:

```
    private Publisher<byte[]> consume(PushbackInputStream pushbackInputStream) {
        // WARNING temporal coupling
        // This method expects to be able to consume and restore inputStream, sizeThresholdInBytes + 1 bytes from pushbackInputStream
```

This will make things more explicit, and we won't add a type leading to a false perception of safety.

Reply:
Have a look at https://github.com/linagora/james-project/commit/41865e606f6c2d7666a64092ac05868857090c1b
It better respects SRP and make main code more readable in my opinion.
Let me know what you think about it